File: backend/app/api/v1/technician_availability.py
from uuid import UUID
from typing import List, Optional
from fastapi import APIRouter, Depends
from app.models import (
    TechnicianAvailabilityCreate,
    TechnicianAvailabilityResponse,
    TechnicianAvailabilityUpdate,
)
from app.services import TechnicianAvailablityService
from app.dependencies import (
    technician_availability_service_dependency,
    get_technician_availability_service,
)

router: APIRouter = APIRouter(
    prefix="/technician_availability", tags=["Technician Availability"]
)


@router.post("/", response_model=TechnicianAvailabilityResponse, status_code=201)
async def create_technician_availability(
    data: TechnicianAvailabilityCreate,
    technician_availability_service: technician_availability_service_dependency,
) -> TechnicianAvailabilityResponse:
    return await technician_availability_service.create_technician_availability(data)


@router.get(
    "/{technician_availability_id}",
    response_model=TechnicianAvailabilityResponse,
    status_code=200,
)
async def readone_technician_availability(
    technician_availability_id: UUID,
    technician_availability_service: technician_availability_service_dependency,
) -> TechnicianAvailabilityResponse:
    return await technician_availability_service.readone_technician_availability(
        technician_availability_id
    )


@router.get("/", response_model=List[TechnicianAvailabilityResponse], status_code=200)
async def readall_technician_availabilities(
    day: Optional[int] = None,
    start_time: Optional[str] = None,
    end_time: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    technician_availability_service: TechnicianAvailablityService = Depends(
        get_technician_availability_service
    ),
) -> List[TechnicianAvailabilityResponse]:
    return await technician_availability_service.readall_technician_availabilities(
        day, start_time, end_time, skip, limit
    )


@router.get(
    "/technician/{technician_id}",
    response_model=List[TechnicianAvailabilityResponse],
    status_code=200,
)
async def readall_technician_availabilities_by_technician_id(
    technician_id: UUID,
    day: Optional[int] = None,
    start_time: Optional[str] = None,
    end_time: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    technician_availability_service: TechnicianAvailablityService = Depends(
        get_technician_availability_service
    ),
) -> List[TechnicianAvailabilityResponse]:
    return await technician_availability_service.readall_technician_availabilities_by_technician_id(
        technician_id, day, start_time, end_time, skip, limit
    )


@router.put(
    "/{technician_availability_id}",
    response_model=TechnicianAvailabilityResponse,
    status_code=200,
)
async def update_technician_availability(
    technician_availability_id: UUID,
    technician_id: UUID,
    data: TechnicianAvailabilityUpdate,
    technician_availability_service: TechnicianAvailablityService = Depends(
        get_technician_availability_service
    ),
) -> TechnicianAvailabilityResponse:
    return await technician_availability_service.update_technician_availability(
        technician_availability_id, technician_id, data
    )


@router.delete("/{technician_availability_id}", response_model=bool, status_code=200)
async def delete_technician_availability(
    technician_availability_id: UUID,
    technician_availability_service: TechnicianAvailablityService = Depends(
        get_technician_availability_service
    ),
) -> bool:
    return await technician_availability_service.delete_technician_availability(
        technician_availability_id
    )

File: frontend/src/lib/services/technician_availability.ts
import axios from "axios";
import { BASE_API_URL } from "@/lib/constants";
import type {
    TechnicianAvailabilityCreate,
    TechnicianAvailabilityUpdate,
    TechnicianAvailabilityResponse,
} from "@/lib/types/technician_availability";

const URL = `${BASE_API_URL}/technician_availability`;

export async function createTechnicianAvailability(data: TechnicianAvailabilityCreate): Promise<TechnicianAvailabilityResponse> {
    const response = await axios.post(URL, data);
    return response.data;
}

export async function readoneTechnicianAvailability(technician_availability_id: string): Promise<TechnicianAvailabilityResponse> {
    const response = await axios.get(`${URL}/${technician_availability_id}`);
    return response.data;
}

export async function readallTechnicianAvailabilities(
    day?: number,
    start_time?: string,
    end_time?: string,
    skip: number = 0,
    limit: number = 100,
): Promise<TechnicianAvailabilityResponse[]> {
    const params: { [key: string]: string | number } = {};

    if (day !== undefined) {
        params.day = day;
    }
    if (start_time !== undefined) {
        params.start_time = start_time;
    }
    if (end_time !== undefined) {
        params.end_time = end_time;
    }

    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(URL, { params });
    return response.data;
}

export async function readallTechnicianAvailabilitiesByTechnician(
    technicianID: string,
    day?: number,
    startTime?: string,
    endTime?: string,
    skip: number = 0,
    limit: number = 100,
): Promise<TechnicianAvailabilityResponse[]> {
    const params: { [key: string]: string | number } = {};

    if (day !== undefined) {
        params.day = day;
    }
    if (startTime !== undefined) {
        params.start_time = startTime;
    }
    if (endTime !== undefined) {
        params.end_time = endTime;
    }

    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(`${URL}/${technicianID}`, { params });
    return response.data;
}

export async function deleteTechnicianAvailability(technician_availability_id: string): Promise<TechnicianAvailabilityResponse> {
    const response = await axios.delete(`${URL}/${technician_availability_id}`);
    return response.data;
}

export async function updateTechnicianAvailability(technician_availability_id: string, data: TechnicianAvailabilityUpdate): Promise<TechnicianAvailabilityResponse> {
    const response = await axios.put(`${URL}/${technician_availability_id}`, data);
    return response.data;
}
